File: Python/106. Construct Binary Tree from Inorder and Postorder Traversal.py
'''
递归
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        def traversal(inorder, postorder, in_begin, in_end, post_begin, post_end):
            if post_begin == post_end:
                return None
            mid = postorder[post_begin:post_end][-1]
            mid_index = 0
            for i in range(in_begin, in_end):
                if inorder[i] == mid:
                    mid_index = i
                    break
            node = TreeNode(mid)
            left_len = mid_index - in_begin
            right_len = in_end - mid_index - 1
            node.left = traversal(inorder, postorder, in_begin, mid_index, post_begin, post_begin + left_len)
            node.right = traversal(inorder, postorder, mid_index + 1, in_end, post_begin + left_len, post_end - 1)
            return node
        return traversal(inorder, postorder, 0, len(inorder), 0, len(postorder))

File: C++/106. Construct Binary Tree from Inorder and Postorder Traversal.cpp
/*
ตÝน้
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* traversal(vector<int>& inorder, vector<int>& postorder) {
        if (postorder.size() == 0) return NULL;
        TreeNode* node = new TreeNode(postorder.back());
        int mid = 0;
        for (int i = 0; i < inorder.size(); i++) {
            if (inorder[i] == postorder.back()) {
                mid = i;
                break;
            }
        }
        vector<int>left_inorder(inorder.begin(), inorder.begin() + mid);
        vector<int>right_inorder(inorder.begin() + mid + 1, inorder.end());
        vector<int>left_postorder(postorder.begin(), postorder.begin() + mid);
        vector<int>right_postorder(postorder.begin() + mid, postorder.end() - 1);

        node->left = traversal(left_inorder, left_postorder);
        node->right = traversal(right_inorder, right_postorder);
        return node;

    }
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return traversal(inorder, postorder);
    }
};